File: sicstus/bin_packing.pl
/*

  Bin packing in SICStus Prolog.

  Simple bin packing problems

  Compare to the following models
  * MiniZinc: http://www.hakank.org/minizinc/bin_packing.mzn
  * ECLiPSe : http://www.hakank.org/eclipse/bin_packing.ecl


  This model is based on the ECLiPSe model but has been
  completely rewritten. 

  Mats Carlsson suggested the solution in bin_packing2/1.

  Model created by Hakan Kjellerstrand, hakank@bonetmail.com
  See also my SICStus Prolog page: http://www.hakank.org/sicstus/

*/

:-use_module(library(clpfd)).
:-use_module(library(lists)).
:-use_module(library(samsort)).


% sorted Stuff
go :-
        ( for(P,0,9) do
              bin_packing(P,sorted)
        ).


% Unsorted Stuff: very slow for problem 8 and 9
go2 :-
        ( for(P,0,9) do
              bin_packing(P,unsorted)
        ).

% Mats Carlsson's variant with unsorted 
go3 :-
        ( for(P,0,9) do
              bin_packing2(P)
        ).


bin_packing(Problem, SortMode) :-

        problem(Problem, StuffUnordered, BinCapacity),

        % sorting (reversed) the stuff makes it much faster
        % (but maybe this is considered as cheating)
        (
            SortMode = sorted 
        -> 
            samsort(#>=,StuffUnordered,Stuff)
        ;
            Stuff = StuffUnordered
        ),

        format('\nProblem ~w\n',[Problem]),

        % BinCapacity: the (common) capacity for each bin
        write(bin_capacity:BinCapacity),nl,

        % Stuff: values/weight of the things to pack
        length(Stuff,NumStuff), 
        write(stuff:Stuff),nl,
        write(num_stuff:NumStuff),nl,

        % Number of bins cannot exceed num_stuff...
        NumBins = NumStuff,

        % Sanity clause: 
        %    No thing can be larger than capacity.
        ( foreach(S,Stuff),
          param(BinCapacity) do
              S > BinCapacity 
        -> 
          format('Stuff ~d is larger than BinCapacity ~d\n', [S, BinCapacity])
        ;
          true
        ),
        
        % Bins: where to put a thing.
        matrix(Bins,[NumBins,NumStuff]),
        append(Bins,BinsList),
        domain(BinsList,0,1),        

        % BinLoads: contains how many things (the summed 
        %           weights of) a bin takes
        length(BinLoads,NumBins),
        domain(BinLoads,0,BinCapacity),
        ( foreach(Bin,Bins),
          foreach(Load,BinLoads),
          param(Stuff) do
              scalar_product(Stuff,Bin,#=,Load)
        ),
       
        % a thing is packed exactly once 
        transpose(Bins,BinsTransposed),
        ( foreach(Column, BinsTransposed) 
        do
            sum(Column,#=, 1)
        ),

        % compare the total loads
        sum(Stuff,#=,StuffSum),
        sum(BinLoads,#=,BinLoadsSum),
        StuffSum #= BinLoadsSum,

        % load the bins in order:
        % first bin must be loaded, and the list must be ordered
        element(1, BinLoads, BinLoads1),
        BinLoads1 #> 0,
        ( fromto(BinLoads,[This,Next|Rest],[Next|Rest],[_])
        do
          This #>= Next
        ),

        % calculate number of loaded bins 
        % (which we will minimize)
        % domain([NumLoadedBins],0,1000000),
        ( foreach(Load2,BinLoads),
          fromto(0,In,Out,NumLoadedBins),
          param(BinCapacity)
        do
          Loaded in 0..1,
          Load2 #> 0 #<=> Loaded #= 1,
          indomain(Loaded),
          Out #= In + Loaded
        ),

        % search
        append(BinsList,BinLoads,Vars),
        % append([NumLoadedBins],Vars1,Vars),
        labeling([ff,step,down,minimize(NumLoadedBins)], Vars),

        % output
        write(bin_loads:BinLoads),nl,
        write(num_loaded_bins:NumLoadedBins),nl,

        % just print smaller problems
        (NumBins =< 40 
        -> 
         % pretty_print(Bins) 
         pretty_print2(Bins) 
        ; 
         write('Bins is too large to print.\n'), true
        ),
        fd_statistics,
        nl.

% This version using cumulative was suggested by Mats Carlsson.
bin_packing2(Problem) :-
        problem(Problem, StuffUnordered, BinCapacity),
        portray_clause(problem(Problem, StuffUnordered, BinCapacity)),
        length(StuffUnordered, N),
        N1 is N+1,
        (   foreach(H,StuffUnordered),
            foreach(task(S,1,E,H,1),Tasks),
            foreach(NH-S,Tagged),
            foreach(S,Assignment),
            foreach(_-X,Keysorted),
            foreach(X,Vars),
            param(N1)
        do  S in 1..N1,
            E in 2..N1,
            NH is -H
        ),
        keysort(Tagged, Keysorted),
        maximum(Max, Vars),
        cumulative(Tasks, [limit(BinCapacity)]),
        labeling([minimize(Max),step], Vars),
        portray_clause(assignment(Assignment)),
        fd_statistics,
        nl.



%
% pretty prints the Bin matrix.

pretty_print(X) :-
        ( foreach(Row,X) 
        do
          sum(Row,#=,Sum),
          ( foreach(R,Row) do
                format('~d ', R)
          ),
          format(" = ~d\n", [Sum])
        ).

% just print the filled bins
pretty_print2(X) :-
        ( foreach(Row,X) 
        do
          sum(Row,#=,Sum),
          Sum > 0 -> 
              (
                  ( foreach(R,Row) do
                        format('~d ', R)
                  ),
                  format(" = ~d\n", [Sum])
              )
        ;
              true
        ).


% Suggested by Mats Carlsson
matrix(_, []) :- !.
matrix(L, [Dim|Dims]) :-
        length(L, Dim),
        (   foreach(X,L),
            param(Dims)
        do  matrix(X, Dims)
        ).


%
% Data
%        

% easy problem to test with
problem(0, 
        [1,2,3,4,5,6,7,8,9,10],
        30).


%
% Example from the Alice system
% Copying files to disks
% http://news.mozart-oz.org/alice/manual/cptutorial/node55.html
%
% """
% Suppose, you want to copy a set of files from your hard-disk onto as 
% few as possible diskettes of a given size, e. g. onto common 1.44 MB 
% diskettes. In case your files do not fit on a single diskette, it might 
% become quite tricky to figure out the minimal number of needed diskettes 
% and how to partition the files.
% """
problem(1, 
        [360, 850, 630, 70, 700, 210], 
        1440).


% simple (and probably unrealistic) packing
problem(2, 
        [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20],
        20).


% simple (and probably even less unrealistic) packing
problem(3, 
        [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25],
        50).


% This problem below is from
% http://www.dcs.gla.ac.uk/~pat/cpM/slides/binPacking.ppt
% 1D Bin Packing (or "CP? Who cares?"), page 3
% and also from
% http://www.dcs.gla.ac.uk/~pat/cpM/JChoco/binPack
%
% num_stuff = 10;
% stuff = [42,63,67,57,93,90,38,36,45,42];
% bin_capacity = 150;
problem(4, 
        [42,69,67,57,93,90,38,36,45,42], 
        150).


% same source of data, but now there are 22 things
problem(5, 
        [42,69,67,57,93,90,38,36,45,42,33,79,27,57,44,84,86,92,
         46,38,85,33],
        250).


%
% continuing of the above example.
%
problem(6, 
        [42,69,67,57,93,90,38,36,45,42,33,79,27,57,44,84,86,92,
         46,38,85,33,82,73,49,70,59,23,57,72,74,69,33,42,28,46,
         30,64,29,74,41,49,55,98,80,32,25,38,82,30], 
        290).


% ibid. 
problem(7,
        [42,69,67,57,93,90,38,36,45,42,33,79,27,57,44,84,86,92,46,38,
         85,33,82,73,49,70,59,23,57,72,74,69,33,42,28,46,30,64,29,74,
         41,49,55,98,80,32,25,38,82,30,35,39,57,84,62,50,55,27,30,36,
         20,78,47,26,45,41,58,98,91,96,73,84,37,93,91,43,73,85,81,79,
         71,80,76,83,41,78,70,23,42,87,43,84,60,55,49,78,73,62,36,44,
         94,69,32,96,70,84,58,78,25,80,58,66,83,24,98,60,42,43,43,
         39],
        500).


% From 
% Graham Kendall: Bin Packing made Easier 
% http://research-reflections.blogspot.com/2009/07/bin-packing-made-easier.html
problem(8,
        [442,252,127,106,37,10,10,252,252,127,106,37,10,9,
        252,252,127,85,12,10,9,252,127,106,84,12,10,252,
        127,106,46,12,10],
        524).


% Variant: remove 46 from the problem above
problem(9,
        [442,252,127,106,37,10,10,252,252,127,106,37,10,9,
        252,252,127,85,12,10,9,252,127,106,84,12,10,252,
        127,106,12,10],
        524).

File: z3/bin_packing.py
#!/usr/bin/python -u
# -*- coding: latin-1 -*-
# 
# in Z3
#
# From Global Constraint Catalogue
# http://www.emn.fr/x-info/sdemasse/gccat/Cbin_packing.html
# """
# Given several items of the collection ITEMS (each of them having a specific 
# weight), and different bins of a fixed capacity, assign each item to a bin 
# so that the total weight of the items in each bin does not exceed CAPACITY.
# 
# Example
#   <(5,<bin-3 weight-4, bin-1 weight-3,bin-3 weight-1>)>
# 
#  The bin_packing constraint holds since the sum of the height of items 
# that are assigned to bins 1 and 3 is respectively equal to 3 and 5. 
# The previous quantities are both less than or equal to the maximum 
# CAPACITY 5. Figure 4.35.1 shows the solution associated with the example.
# 
# Remark
# 
# Note the difference with the classical bin-packing problem [MT90] where 
# one wants to find solutions that minimise the number of bins. In our 
# case each item may be assigned only to specific bins (i.e., the different 
# values of the bin variable) and the goal is to find a feasible solution. 
# This constraint can be seen as a special case of the cumulative 
# constraint [AB93], where all task durations are equal to 1.
# """
# 
# This Z3 model was written by Hakan Kjellerstrand (hakank@gmail.com)
# See also my Z3 page: http://hakank.org/z3/
# 
from z3_utils_hakank import *


#
# bin_packing
#
# Note: capacity (and bins) might be IntVar but weights must be an int vector
#
def bin_packing(sol,capacity, bins, weights):
  n = len(bins)
  for b in range(n):
    sol.add(Sum([ weights[j]*If(bins[j] == b,1,0) for j in range(n)] ) <= capacity)


#
# print the packing
#
def print_packing(weights,bins):
    n = len(weights)
    for i in range(n):
      total = 0
      packed = []
      for j in range(n):
        if bins[j] == i:
          total += weights[j]
          packed.append(j)
      if total > 0:
        print "bin %2i"%i , " items:",  packed, " total weight:", total
    print

#
# minimize the capacity (if capacity1 is not set)
#
def test_bin_packing_min_capacity(weights,capacity1=None):

  sol = Solver()

  n = len(weights)

  for i in range(n):
    assert weights[i] <= capacity, "All weights must be <= capacity: "

  # variables
  bins = makeIntVector(sol,"bins",n,0,n-1)

  capacity = Int("capacity")
  if capacity1 != None:
    sol.add(capacity == capacity1)
  else:
    sol.add(capacity >= max(weights))

  # constraints
  bin_packing(sol, capacity, bins, weights)

  num_solutions = 0
  while sol.check() == sat:
    num_solutions += 1
    mod = sol.model()
    print "weights:", weights, "capacity  :", mod.eval(capacity)
    print "bins :", [mod.eval(bins[i]) for i in range(n)]
    bbins = [mod.eval(bins[i]) for i in range(n)]
    print_packing(weights,bbins)
    
    if capacity1 == None:
      getLessSolution(sol,mod,capacity)
    else:
      getDifferentSolution(sol,mod,bins)

  print "num_solutions:", num_solutions  

#
# Minimize the number of bins
# Here both weights and capacity must be ints
#
def test_bin_packing_min_bins(weights,capacity):

  sol = Solver()
  # sol = Optimize()

  n = len(weights)
  # for i in range(n):
  #  assert weights[i] <= capacity, "All weights must be <= capacity: "



  # variables
  bins = makeIntVector(sol,"bins",n,0,n-1)
  weight_per_bin = makeIntVector(sol,"weight_per_bin", n, 0, capacity)

  max_bin_used = makeIntVar(sol,"max_bin_used", 0,n-1)

  # constraints
  
  # find the max number of bin used
  maximum(sol,max_bin_used,bins)

  # bin_packing(sol, capacity, bins, weights)
  for b in range(n):
    # this includes the bin_packing constraint
    sol.add(weight_per_bin[b] == Sum([ weights[j]*If(bins[j] == b,1,0) for j in range(n)] ))
    sol.add(weight_per_bin[b] <= capacity)

  decreasing(sol,weight_per_bin)

  # sol.minimize(max_bin_used)
  num_solutions = 0
  while sol.check() == sat:
  # if sol.check() == sat:      
    num_solutions += 1
    mod = sol.model()
    print "weights:", weights
    print "capacity  :", capacity
    print "max_bin_used:", mod.eval(max_bin_used)
    print "weight_per_bin:", [mod.eval(weight_per_bin[i]) for i in range(n)]
    bbins = [mod.eval(bins[i]) for i in range(n)]
    print "bins :", bbins
    print_packing(weights,bbins)
    print
    getLessSolution(sol,mod,max_bin_used)

  print "num_solutions:", num_solutions  


# n = 3
# weights = [4,3,1]
# capacity = 4

# simple (and probably unrealistic) packing
# capacity = 20
weights = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20];
capacity = 30


# The problem below is from
# http://www.dcs.gla.ac.uk/~pat/cpM/slides/binPacking.ppt
# 1D Bin Packing (or "CP? Who cares?"), page 3
# and also from
# http://www.dcs.gla.ac.uk/~pat/cpM/JChoco/binPack
#
# num_stuff = 10;
weights = [42,63,67,57,93,90,38,36,45,42]
capacity = 150;
# capacity = 100;


# This problem instance is via
# http://yetanothermathprogrammingconsultant.blogspot.com/2011/08/bin-packing.html
# and is "critical" to the result in
# "Bin Packing, What is It?"
# http://www.developerfusion.com/article/5540/bin-packing/
# where the minimum bins are 19,
# whereas Erwin K's code have 17 as the optimal result.
# 
# LazyFD solves this in 1:44 minutes (i.e. with result 17) 
# when num_bins is hardcoded to 19 (as in Erwin's model)
# 
weights = [
 26,   57, 18,   8,  45,
 16,   22, 29,   5,  11,
  8,   27, 54,  13,  17,
 21,   63, 14,  16,  45,
  6,   32, 57,  24,  18,
 27,   54, 35,  12,  43,
 36,   72, 14,  28,   3,
 11,   46, 27,  42,  59,
 26,   41, 15,  41,  68]
capacity = 80

# same source of data, but now there are 22 things
# num_stuff = 22; % number of things
# weights = [42,69,67,57,93,90,38,36,45,42,33,79,27,57,44,84,86,92,46,38,85,33]
# capacity = 250

# still more stuff
# weights = [42,69,67,57,93,90,38,36,45,42,33,79,27,57,44,84,86,92,46,38,85,33,82,73,49,70,59,23,57,72,74,69,33,42,28,46,30,64,29,74,41,49,55,98,80,32,25,38,82,30] # 35,39,57,84,62,50,55,27,30,36,20,78,47,26,45,41,58,98,91,96,73,84,37,93,91,43,73,85,81,79,71,80,76,83,41,78,70,23,42,87,43,84,60,55,49,78,73,62,36,44,94,69,32,96,70,84,58,78,25,80,58,66,83,24,98,60,42,43,43,3]
# capacity = 290

# From 
# Graham Kendall: Bin Packing made Easier 
# http://research-reflections.blogspot.com/2009/07/bin-packing-made-easier.html
weights = [442,252,127,106,37,10,10,252,252,127,106,37,10,9,
           252,252,127,85,12,10,9,252,127,106,84,12,10,252,127,106,46,12,10]
capacity = 524


# # Variant: remove 46 from the problem above
# weights = [442,252,127,106,37,10,10,252,252,127,106,37,10,9,
#            252,252,127,85,12,10,9,252,127,106,84,12,10,252,
#            127,106,12,10]
# capacity = 524



# test_bin_packing(weights)
# test_bin_packing(weights,capacity)
test_bin_packing_min_bins(weights,capacity)
